Changes for Pyfluent UI. (#870)
pyansys/pyfluent: doc/source/api/core/solver/fielddata.rst

Field data
==========

> You can use field data objects to access Fluent surface, scalar, vector and pathlines data.

Reviewer: ```suggestion
You can use field data objects to access Fluent surface, scalar, vector, and pathline data.
```

Simple requests
---------------

> Following are the separate methods for requesting each type of field:

Reviewer: ```suggestion
Here are the methods for requesting each type of field:
```

- ``get_surface_data`` for surface data.
- ``get_scalar_field_data`` for scalar field data.
- ``get_vector_field_data`` for vector field data
- ``get_pathlines_field_data`` for vector field data

> get_surface_data

Reviewer: ```suggestion
Get surface data
```

~~~~~~~~~~~~~~~~

> You can request surface vertices for a given ``surface_name`` by calling
> ``get_surface_data`` and specifying ``Vertices`` for the data_type.

Reviewer: ```suggestion
You can request surface vertices for a given ``surface_name`` by calling
the ``get_surface_data`` method and specifying ``Vertices`` for ``data_type``.
```

.. code-block:: python

  >>> from ansys.fluent.core.services.field_data import SurfaceDataType

  >>> field_data.get_surface_data(surface_name="inlet", data_type=SurfaceDataType.Vertices)
  {4: array([-0.34760258,  0.        , -0.04240644, ..., -0.2953132 ,
        0.        , -0.06207064], dtype=float32)}

> You can call the same method to get the corresponding surface face-connectivity, normals and centroids

Reviewer: ```suggestion
You can call the same method to get the corresponding surface face connectivity, normals, and centroids
```

> by specifying ``FacesConnectivity``, ``FacesNormal`` and ``FacesCentroid`` respectively for the data_type.

Reviewer: ```suggestion
by specifying ``FacesConnectivity``, ``FacesNormal`` and ``FacesCentroid`` respectively for ``data_type``.
```

.. code-block:: python

  >>> field_data.get_surface_data(surface_name="inlet", data_type=SurfaceDataType.FacesConnectivity)
  {4: array([  4,   3,   2, ..., 379, 382, 388])}

  >>> field_data.get_surface_data(surface_name="inlet", data_type=SurfaceDataType.FacesNormal)
  {4: array([ 0.0000000e+00,  2.5835081e-06,  ...,
          2.7459005e-06,  0.0000000e+00,  0.0000000e+00,  3.0340884e-06], dtype=float32)}
  
  >>> field_data.get_surface_data(surface_name="inlet", data_type=SurfaceDataType.FacesCentroid)
  {4: array([-0.34724122,  0.        , -0.0442204 , -0.34724477, ...,
         -0.04050309, -0.34645557,  0.        , -0.04421487, -0.34646705], dtype=float32)}
       
Response contains a dictionary of surface IDs to numpy array of the requested field. 

> get_scalar_field_data

Reviewer: ```suggestion
Get scalar field data
```

~~~~~~~~~~~~~~~~~~~~~

> You can call ``get_scalar_field_data`` to get scalar field data, such as temperature:

Reviewer: ```suggestion
You can call the ``get_scalar_field_data`` method to get scalar field data, such as temperature:
```

.. code-block:: python

  >>> field_data.get_scalar_field_data(surface_name="inlet", field_name="temperature")
  {4: array([922.6778 , 923.7954 , 923.791  , 922.67017, 922.88135,  ...,
         924.94556, 924.9451 , 924.9536 , 923.87366, 922.8334 , 922.82434], dtype=float32)}
         

> Response contains a dictionary of surface IDs to numpy array of the requested field.          

Reviewer: ```suggestion
The response contains a dictionary of surface IDs to a numpy array of the requested field.          
```

> get_vector_field_data

Reviewer: ```suggestion
Get vector field data
```

~~~~~~~~~~~~~~~~~~~~~

> You can call ``get_vector_field_data`` to get vector field data. 

Reviewer: ```suggestion
You can call the ``get_vector_field_data`` method to get vector field data. 
```

.. code-block:: python

  >>> field_data.get_vector_field_data(surface_name="inlet", vector_field="velocity")
  {4: (array([ 5.81938386e-01, -1.01187916e+01,  3.14891455e-03,  ...,
        -6.49216697e-02,  1.44923580e+00, -1.04387817e+01], dtype=float32), 0.00012235096)}
  
Response is a dictionary of surface IDs to a tuple containing a numpy array of ``vector field`` and ``vector-scale``. 

> get_pathlines_field_data

Reviewer: ```suggestion
Get pathlines field data
```

~~~~~~~~~~~~~~~~~~~~~~~~  

> You can call ``get_pathlines_field_data`` to get pathlines data. 

Reviewer: ```suggestion
You can call the ``get_pathlines_field_data`` method to get pathlines field data. 
```

Reviewer: You need to add the word "pathlines" to the Vale accept list. When Vale is implemented for the repo, it should be here: C:\AnsysDev\GitRepos\PyAnsys\pyfluent-ui\doc\styles\Vocab\ANSYS\accept.txt. However, I pulled from "main" and am not seeing it?!

.. code-block:: python

  >>> field_data.get_pathlines_field_data(surface_name="inlet", field_name="temperature")
  {4: {'faces': array([    2,     0,     1, ...,     2, 28828, 28829]), 'vertices': array([-0.34724122,  0.        , -0.0442204 , ..., -0.19877256,
         -0.1250188 , -0.0621446 ], dtype=float32), 'temperature': array([879.25073, 828.8636 , 861.65686, ..., 892.1276 , 894.6389 ,
         897.95325], dtype=float32)}}

> Pathlines data is returned as line surface. So response is a dictionary of surface IDs to a information about line surface. 

Reviewer: We should be consistent in terminology. Is it simply "pathlines data" or "pathlines field data"?

.. note:: 
   In Fluent, surface name can be associated with multiple surface IDs.
   Thus, a response contain a surface ID as a key of the returned dictionary. 

Making multiple requests in a single transaction
------------------------------------------------

> you can get data for multiple fields in a single request and receive all the requested data 

Reviewer: ```suggestion
You can get data for multiple fields in a single request and receive all the requested data 
```

in a single response.

The ``add_get_<items>_request`` methods combine requests for multiple fields in a single request: 

- ``add_get_surfaces_request`` adds a surfaces request.
- ``add_get_scalar_fields_request`` adds a scalar fields request.
- ``add_get_vector_fields_request`` adds a vector fields request. 
- ``add_get_pathlines_fields_request`` adds a pathlines fields request.

> You can build a request by making multiple calls on the field data object, as follows: 

Reviewer: ```suggestion
You can build a request by making multiple calls on the field data object, as follows:
```

.. code-block::

>>> field_data.add_get_surfaces_request(surface_ids=[1], provide_vertices=True, 
                                        provide_faces=False, provide_faces_centroid=True
                                       )
>>> field_data.add_get_surfaces_request(surface_ids=[2], provide_vertices=True, 
                                       provide_faces=True
                                       )
>>> field_data.add_get_scalar_fields_request(surface_ids=[1,2], field_name="temperature",
                                            node_value=True, boundary_value=True
                                            )                                            
>>> field_data.add_get_vector_fields_request(surface_ids=[1,2])
>>> field_data.add_get_pathlines_fields_request(surface_ids=[1,2], field_name="temperature"                                           
                                            )                                            

                                            

> You can call ``get_fields`` to get the data for all these requests. That call also

Reviewer: ```suggestion
You can call the ``get_fields`` method to get the data for all these requests. This call also
```

> clears all requests, so that subsequent calls to ``get_fields`` yield nothing until

Reviewer: ```suggestion
clears all requests, so that subsequent calls to the ``get_fields`` method yield nothing until
```

>>> payload_data = field_data.get_fields()

``payload_data`` is a dictionary containing the requested fields as a numpy array in the following order:

``tag [Tuple]-> surface_id [int] -> field_name [str] -> field_data[np.array]``

  
Tag
--- 

> Fluent version < 23.1
> ~~~~~~~~~~~~~~~~~~~~~ 

Reviewer: ```suggestion
Fluent version earlier than 23.1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

> A tag ID is generated by applying ```bitwise or``` on all tags for a request. Here is a list

Reviewer: Should the format be ``"bitwise or"``? I've never seen three back ticks used.

- ``add_get_surfaces_request`` adds a surfaces request.

> - ``add_get_scalar_fields_request`` adds a scalar fields request.

Author: @PipKat  Kindly review the changes in this file. Thanks

Reviewer: @ajain-work Sorry but I was out of commission for several days due to Hurricane Ian here in Florida. I'll take a look at your Vale errors.

Author: @PipKat Thanks a lot for reviewing and providing feedback.

*  BOUNDARY_VALUES: 8,

For example, if you request the scalar field data for element location[2], in the
dictionary,  ``tag`` is ``2``. Similarly, if you request the boundary values[8] for 
node location[4], ``tag`` is ``(4|8)`` or 12.

> Fluent version >= 23.1
> ~~~~~~~~~~~~~~~~~~~~~~ 

Reviewer: ```suggestion
Fluent versions 23.1 and later
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

A tag is tuple of input, value pairs for which field data is generated.

For example, if you request the scalar field data for element location, in the
dictionary,  ``tag`` is ``(('type','scalar-field'), ('dataLocation', 1), ('boundaryValues',False))``. 
Similarly, if you request the boundary values for node location, ``tag`` is 

> ``(('type','scalar-field'), ('dataLocation', 0), ('boundaryValues',True)``

Reviewer: ```suggestion
``(('type','scalar-field'), ('dataLocation', 0), ('boundaryValues',True)``.
```

- ``vector field``, with the same name as the vector field name that is passed in the request 
- ``vector-scale``, a float value indicating the vector scale.
 
Pathlines field request
~~~~~~~~~~~~~~~~~~~~~~~

> The response to a pathlines field request contains following fields:

Reviewer: ```suggestion
The response to a pathlines field request contains the following fields:
```

- ``pathlines-count``, pathlines count. 

> For each pathline following fields are returned.

Reviewer: ```suggestion
For each pathline, the following fields are returned:
```
